app/controllers/application_controller.rb

```diff
@@ -1,10 +1,145 @@
+require 'pry'
+
 require './config/environment'
 
+
 class ApplicationController < Sinatra::Base
 
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    enable :sessions
+    set :session_secret, "secret"
+  end
+
+  get '/' do 
+    erb :index
+  end
+
+  get '/signup' do 
+    if logged_in?
+      redirect '/tweets'
+    else
+      erb :'/users/create_user'
+    end
+  end
+
+  post '/signup' do 
+    @user = User.create(username: params[:username], password: params[:password], email: params[:email])
+    if params[:username] == "" || params[:password] == "" || params[:email] == "" 
+      redirect '/signup'
+    else
+      session[:id] = @user.id
+      redirect '/tweets' 
+    end
+  end
+
+
+  get '/login' do 
+    if logged_in?
+      redirect '/tweets'
+    else
+      erb :'/users/login'
+    end
+  end
+
+  post '/login' do 
+    user = User.find_by(:username => params[:username])
+    if user && user.authenticate(params[:password])
+      session[:id] = user.id
+      redirect '/tweets'
+    else
+      redirect '/login' 
+    end
+  end 
+
+  get '/tweets/new' do 
+
+    if logged_in?
+      erb :'/tweets/create_tweet'
+    else
+      redirect '/login'
+    end
   end
 
-end
\ No newline at end of file
+  post '/tweets/new' do   
+    if params[:content] == "" 
+      redirect '/tweets/new'
+    else
+      @tweet = Tweet.new(content: params[:content])  
+      @user = current_user
+      @tweet.user_id = @user.id 
+      @tweet.save
+      redirect '/tweets'
+    end
+   end
+
+  get '/tweets' do 
+    if logged_in?
+      @user = current_user 
+      erb :'/tweets/tweets'
+    else 
+      redirect '/login'
+    end  
+  end
+
+  get '/users/:slug' do 
+    @user = User.find_by(username: params[:slug])
+    erb :'/tweets/tweets'
+  end
+
+  get '/tweets/:id' do
+    @tweet = Tweet.find_by(id: params[:id])
+    if logged_in?
+      erb :'/tweets/show_tweet' 
+    else
+      redirect '/login'
+    end
+  end
+
+  get '/tweets/:id/edit' do 
+    @tweet = Tweet.find_by(id: params[:id])
+    if logged_in?
+      erb :'/tweets/edit_tweet'
+    else
+      redirect '/login'
+    end
+
+  end
+
+  patch '/tweets/:id/edit' do
+     @tweet = Tweet.find_by(id: params[:id])
+     if @tweet.update(content: params[:content])
+       redirect to "/tweets/#{@tweet.id}/edit"
+     else
+       redirect to "/tweets/#{@tweet.id}/edit"
+     end
+   end
+
+  delete '/tweets/:id/delete' do
+    if logged_in?
+       @tweet = Tweet.find_by_id(params[:id])
+       if session[:id] == @tweet.user_id
```

Comment:
Prefer to find tweets that are already only associated to the `current_user` than to manually check the `user_id` foreign key of the tweet against the `current_user.id`

```ruby
tweet = current_user.tweets.find_by(:id => params[:id])
if tweet && tweet.destroy
  redirect "/tweets"
else
  redirect "/tweets/#{tweet.id}"
end
```

We're using the association collection method provided by [`has_many`](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many) to query the user's tweets. 

Note the usage of [`find_by`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find-by) vs [`find`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find). Had we used `find` here, ActiveRecord would raise an error in the event that we could not find a tweet associated with this user. `find_by` will allow this action to fail gracefully, simply returning nil if the user is trying to delete a tweet that they do not have access to.